### PR TITLE
feat: create new validator address (CLI and GUI)

### LINKED ISSUE #713: Inability to Create Validator Address in CLI and GUI Applications
## Describe the problem you'd like to have solved

Users are unable to create a new address for a validator, neither in the CLI nor GUI application. The `address new` command in the CLI only creates an account address.

### Describe the ideal solution

We should add a new optional flag, such as "--address_type," with values like "validator" or "bls_account." The default value should be set to "bls_account" for now.

### Alternatives and current workarounds

Currently, there is no workaround available for this issue.

---

Please close this issue when both the CLI and GUI support creating a new address for validators.

### PR DESCRIPTION
**Pull request by jemiluv8**
## Description
Add ability to Create Validator Address in CLI and GUI Applications

## Related issue(s)
- Fixes #713
/claim #713

## GUI SCREENSHOT
Screenshot 2023-10-20 at 1 36 32 AM

### CLI SCREENSHOT - DEBUGGING AND TESTING
Screenshot 2023-10-20 at 2 08 14 AM

### CODE REVIEW

File: cmd/gtk/assets/ui/widget_wallet.ui [hunk 1]
```diff
@@ -253,6 +253,21 @@
                     <property name="homogeneous">True</property>
                   </packing>
                 </child>
+                <child>
+                  <object class="GtkToolButton" id="id_button_new_validator_address">
```
Comment (jemiluv8) [inline cmd/gtk/assets/ui/widget_wallet.ui:257]: New button to "Add New Validator Address"

File: cmd/gtk/model_wallet.go [hunk 2]
```diff
@@ -85,8 +85,16 @@ func (model *walletModel) rebuildModel() {
 	}()
 }
 
-func (model *walletModel) createAddress() error {
-	address, err := model.wallet.NewBLSAccountAddress("")
+func (model *walletModel) createAddress(accountType wallet.AddressType) error {
```
Comment (jemiluv8) [inline cmd/gtk/model_wallet.go:88]: Modified original createAddress to now take an argument.

File: cmd/gtk/widget_wallet.go [hunk 3]
```diff
@@ -55,6 +56,7 @@ func buildWidgetWallet(model *walletModel) (*widgetWallet, error) {
 	labelEncrypted := getLabelObj(builder, "id_label_wallet_encrypted")
 
 	getToolButtonObj(builder, "id_button_new_address").SetIconWidget(AddIcon())
+	getToolButtonObj(builder, "id_button_new_validator_address").SetIconWidget(AddIcon())
```
Comment (jemiluv8) [inline cmd/gtk/widget_wallet.go:59]: using same AddIcon here

File: cmd/gtk/widget_wallet.go [hunk 4]
```diff
@@ -136,9 +138,10 @@ func buildWidgetWallet(model *walletModel) (*widgetWallet, error) {
 		})
 
 	signals := map[string]interface{}{
-		"on_new_address":     w.onNewAddress,
-		"on_change_password": w.onChangePassword,
-		"on_show_seed":       w.onShowSeed,
+		"on_new_address":           w.onNewAddress,
+		"on_new_validator_address": w.onNewValidatorAddress,
```
Comment (jemiluv8) [inline cmd/gtk/widget_wallet.go:142]: Introduce new singnal(on_new_validator_address) for creating new validator address

File: cmd/gtk/widget_wallet.go [hunk 5]
```diff
@@ -148,7 +151,12 @@ func buildWidgetWallet(model *walletModel) (*widgetWallet, error) {
 }
 
 func (ww *widgetWallet) onNewAddress() {
-	err := ww.model.createAddress()
+	err := ww.model.createAddress(wallet.AddressTypeBLSAccount)
```
Comment (jemiluv8) [inline cmd/gtk/widget_wallet.go:154]: Updated this to specify address type

File: cmd/gtk/widget_wallet.go [hunk 6]
```diff
@@ -148,7 +151,12 @@ func buildWidgetWallet(model *walletModel) (*widgetWallet, error) {
...
+	errorCheck(err)
+}
+
+func (ww *widgetWallet) onNewValidatorAddress() {
```
Comment (jemiluv8) [inline cmd/gtk/widget_wallet.go:158]: Call back for 'on_new_validator_address' signal. Creates a validator address.

File: cmd/wallet/address.go [hunk 7]
```diff
@@ -76,12 +77,33 @@ func buildNewAddressCmd(parentCmd *cobra.Command) {
...
 	newAddressCmd.Run = func(_ *cobra.Command, _ []string) {
+		var addr string
+		var err error
+
 		label := cmd.PromptInput("Label")
 		wallet, err := openWallet()
 		cmd.FatalErrorCheck(err)
 
-		addr, err := wallet.NewBLSAccountAddress(label)
+		isForBLSAccount := *addressType == w.AddressTypeBLSAccount.String()
```
Comment (jemiluv8) [inline cmd/wallet/address.go:91]: Cli related changes to support specifying command to create validator address

File: wallet/client.go [hunk 8]
```diff
@@ -12,6 +12,17 @@ import (
 	"google.golang.org/grpc/credentials/insecure"
 )
 
+type AddressType string
```
Comment (jemiluv8) [inline wallet/client.go:15]: Introduced AddressType to be used in other places so we don't use the strings directly.

**Review by jemiluv8 — COMMENTED** [review]
@b00f, I just completed a self-review.
This PR is ready for review and any requests for changes.

**Comment by themantre** [comment]
It's better to add a new dialog for creating new address, like:
```
+----------------------------------------------------------------+
| New address dialog                                             |
+----------------------------------------------------------------+
|                                                                |
| Label: [                               ]                       |
| Type:  [   \ Account               \/  ]                       |
|            \ Validator                                         |
|                                                                |
|                                              [Close] [ OK]     |
+----------------------------------------------------------------+
```

File: cmd/wallet/address.go [hunk 9]
```diff
@@ -76,12 +77,33 @@ func buildNewAddressCmd(parentCmd *cobra.Command) {
...
+		isForValidatorAccount := *addressType == w.AddressTypeValidator.String()
+
+		if !isForBLSAccount || !isForValidatorAccount {
```
Comment (themantre) [inline cmd/wallet/address.go:94]: It is complex. You can make it simple like:

if addrType == "fo" { ... }
else if addrType = "bar" { .. }
else { error }

Feel free to use switch/case

**Review by themantre — CHANGES_REQUESTED** [review]

**Comment by jemiluv8** [comment]
ok @themantre, I'm working on the changes.

**Comment by themantre** [comment]
Good job @jemiluv8 . There are some small linting issues.
Please give me time, I will review the code tomorrow.

**Comment by jemiluv8** [comment]
@themantre  I will fix the lint issues. I'm having challenge "refreshing" the
list of addresses after creating a new address from the modal.

Newly created addresses do not show in the addresses table until I switch views
and come back. I'd appreciate any pointers in that regard. 
I'm looking into it in the meantime.

**Comment by jemiluv8** [comment]
Fixed issues with newly created address not appearing in the list of addresses.
Also attempted a fix for lint issues. Its ready for review.

**Comment by themantre** [comment]
@jemiluv8 
Please check https://github.com/jemiluv8/pactus/pull/1

**Review by b00f — APPROVED** [review]

**Comment by b00f** [comment]
@jemiluv8 Thanks

**Pull request merged**